### PR TITLE
Timeout API rework

### PR DESCRIPTION
**Pull request by andyross**
Big (but less so than the first version) API rework to make timeouts an opaque type, allowing for  their specification in alternate representations like 64 bit types, higher precision units and absolute/uptime values.

See #21305 for the RFC detailing the changes.

### CODE REVIEW

File: include/sys_clock.h
Comment (dcpleung) [inline include/sys_clock.h]: Since `k_ticks_t` is unsigned, `K_TICKS_FOREVER` is now of max integer value, instead of actual -1 (in old behavior). Would this affect the comparisons with `K_FOREVER`/`Z_FOREVER` when legacy timeout API is enabled?

Comment (andyross) [reply include/sys_clock.h]: Not if everything is using the macros correctly, which is a compile error when legacy is not enabled.  The decision to make the 32 bit type unsigned and leave it that way in legacy mode was a compromise from the first version of this PR that I didn't change here.  It could be revisited I guess, but I fear this sort of thing can become a bikeshed.

File: drivers/led/lp5562.c
Comment (andrewboie) [inline drivers/led/lp5562.c]: Huh, I was surprised to see these K_MSEC() removed in this C file, were they being applied twice?

File: include/sys_clock.h
Comment (andrewboie) [inline include/sys_clock.h]: I think it would be better to document this function and add it to our APIs alongside where k_sleep() is defined (which IIRC is in kernel.h). at minimum I think doxygen is needed.

File: samples/basic/threads/src/main.c
Comment (andrewboie) [inline samples/basic/threads/src/main.c]: Doesn't this eventually get propagated to k_thread_create()? Why does this have to be a 0 and not K_NO_WAIT?

File: lib/cmsis_rtos_v1/Kconfig
Comment (andrewboie) [inline lib/cmsis_rtos_v1/Kconfig]: we should probably file an enhancement GH ticket that cmsis v1/v2 and POSIX currently forces the legacy API

edit: net buffers and console as well. how much effort is it to convert these?

File: tests/kernel/mem_protect/futex/prj.conf
Comment (andrewboie) [inline tests/kernel/mem_protect/futex/prj.conf]: why does the futex test need this?

File: include/sys_clock.h
Comment (andrewboie) [inline include/sys_clock.h]: this also needs doxygen

File: drivers/led/lp5562.c
Comment (andyross) [reply drivers/led/lp5562.c]: More likely they were being used as terms in an expression (like "take half the blink period").  In general I tried to choose changes to minimize the size of the patch.  So if you had one timeout defined as a number but used unmodified in several of places, I'd put the macro in the declaration.  If it was a time value used with modification, I'd leave it as an integer in the header and change the usage to k_msleep(), etc...

File: include/sys_clock.h
Comment (andyross) [reply include/sys_clock.h]: It gets that in an upcoming patch in the series, where it becomes a real function (because k_sleep() now takes a k_timeout_t).  I put the placeholders here just to grease the refactoring so everything bisects without weird splitting of features.

File: samples/basic/threads/src/main.c
Comment (andyross) [reply samples/basic/threads/src/main.c]: The timeout constructor macros are written in terms of the time unit conversion API, which are C functions.  C function calls, even pure ones that can be evaluated at compile time based on constant arguments, can't be used as a rvalue to a static initializer (C++ has constexpr for this, C never bothered).

We'd need a separate API to construct these, when it's really not that useful (the only use so far isn't to start a thread on a timer, it's to not start a thread at all).  The decision from the last time around was that we could address this with a new thread definition API (e.g. a "K_NO_START" flag).

File: tests/kernel/mem_protect/futex/prj.conf
Comment (andyross) [reply tests/kernel/mem_protect/futex/prj.conf]: It passes its own timeout value around via pointer arguments while assuming it's an integer, and would have been a minor port to the logic.

File: lib/cmsis_rtos_v1/Kconfig
Comment (andyross) [reply lib/cmsis_rtos_v1/Kconfig]: Not a lot, really.  But it involves a little bit of design decision (i.e. does the subsystem want to expose k_timeout_t in its own API or stick with milliseconds and convert, etc..).  One of the sticking points last time was adding changes to subsystems without maintainer approval, so this one limits itself to subsystems that port with only trivial changes.

**Comment by andyross** [comment]
Update to include k_uptime_ticks(), which was in the RFC but been forgotten.  Also sneak in the extra test code for absolute timeouts which got dropped in the first version.

File: lib/cmsis_rtos_v1/Kconfig
Comment (andrewboie) [reply lib/cmsis_rtos_v1/Kconfig]: OK. I think  we should have GH tickets just so we can track it.

**Review by tyrantlizard77 — APPROVED** [review]

**Review by dcpleung — APPROVED** [review]

File: include/kernel.h
Comment (nordic-krch) [inline include/kernel.h]: im missing a place where timeout argument is well described. `k_timeout_t` in `sys_clock.h` is missing description at all and in the api it is very limited. I know that it is because of two modes but still maybe there should be something like: `see k_timeout_t for more details`.

File: include/kernel.h
Comment (nordic-krch) [inline include/kernel.h]: parameter should not be called `ms` anymore.

File: include/kernel.h
Comment (nordic-krch) [inline include/kernel.h]: "in system ticks" would be better.

File: include/kernel.h
Comment (nordic-krch) [inline include/kernel.h]: "in system ticks"

File: include/kernel.h
Comment (nordic-krch) [inline include/kernel.h]: shouldn't it be `from ticks` or `from absolute/uptime given in ticks`?

File: include/kernel.h
Comment (nordic-krch) [inline include/kernel.h]: shouldn't it be `from ms`  or `from absolute/uptime given in ms`?

File: include/kernel.h
Comment (nordic-krch) [inline include/kernel.h]: same concern as above

File: include/kernel.h
Comment (nordic-krch) [inline include/kernel.h]: why not `u64_t`?

File: include/kernel.h
Comment (nordic-krch) [inline include/kernel.h]: in future that should be changed to u64_t as `k_ticks_to_ms_floor64` returns that anyway.

File: include/sys_clock.h
Comment (nordic-krch) [inline include/sys_clock.h]: i'm missing here a common description of `k_timeout_t`. What it represents, what are the special values, etc. Recommendations (like using K_MSEC). This could be referenced by other API's

File: include/sys_clock.h
Comment (nordic-krch) [inline include/sys_clock.h]: should that be just `(k_timeout_t)t`?

File: include/sys_clock.h
Comment (nordic-krch) [inline include/sys_clock.h]: maybe something with `GET_ABS` in name. Otherwise it can be confused with converting t to absolute (positive) value.

File: include/sys_clock.h
Comment (nordic-krch) [inline include/sys_clock.h]: what does this function do? Missing dox.

File: kernel/queue.c
Comment (nordic-krch) [inline kernel/queue.c]: can you explain why loop is no longer needed?

File: kernel/sched.c
Comment (nordic-krch) [inline kernel/sched.c]: there is some inconsistency there: `z_add_thread_timeout` takes `k_timeout_t` as the second argument. In legacy mode that means milliseconds and yet it seems that it takes ticks since ms to ticks conversion occurs.

Why not do that conversion for legacy mode in `z_add_timeout`?

File: kernel/timeout.c
Comment (nordic-krch) [inline kernel/timeout.c]: could it take `k_ticks_t` as argument?

Comment (nordic-krch) [inline kernel/timeout.c]: later on local variable `struct _timeout *t;` is created. Those names should not conflict.

File: kernel/timeout.c
Comment (nordic-krch) [inline kernel/timeout.c]: any reason for removing lock? does not look like related to PR main goal.

File: kernel/Kconfig
Comment (andyross) [reply kernel/Kconfig]: Fair enough.  Though it needs both, since depends won't set it to y.

File: drivers/i2s/Kconfig
Comment (andyross) [reply drivers/i2s/Kconfig]: There are some docs in the commit message.  Broadly subsystems that expose their own timeouts via APIs (e.g. they have functions that can accept K_FOREVER, etc...) need to decide how to do that -- they can either keep their current APIs as millisecond counts, or expose k_timeout_t directly (the latter requires a bit more work).  Andrew asked for GH issues to be filed for each of these, which I'll do.

File: kernel/thread.c
Comment (andyross) [reply kernel/thread.c]: It now happens  at a single place in z_add_timeout().

File: tests/boards/intel_s1000_crb/src/dma_test.c
Comment (andyross) [reply tests/boards/intel_s1000_crb/src/dma_test.c]: Both are valid APIs.  You can use k_sleep() with a proper timeout value, or call k_msleep()/k_usleep() with an integer value you calculate yourself.    Generally I tried to refactor legacy code in a way that naturally matches what they were doing (which is to say: I mostly tried to optimize the line count of the patch).

File: tests/kernel/mem_protect/futex/src/main.c
Comment (andyross) [reply tests/kernel/mem_protect/futex/src/main.c]: Indeed, will fix.

File: samples/scheduler/metairq_dispatch/src/msgdev.c
Comment (andyross) [reply samples/scheduler/metairq_dispatch/src/msgdev.c]: Not understanding.  Yes, of course, but nothing in the legacy API uses that macro, it's new.  So why bother supporting it in legacy mode?

File: kernel/timer.c
Comment (andyross) [reply kernel/timer.c]: There is some documentation in the Z_TICK_ABS() macro explaining the convention.  I can add the convenience wrapper if you like.

File: include/kernel.h
Comment (andyross) [reply include/kernel.h]: You're right, there should be a doc block on k_timeout_t.  Will add one.

File: include/kernel.h
Comment (andyross) [reply include/kernel.h]: Good point.

File: include/kernel.h
Comment (andyross) [reply include/kernel.h]: Meh.  Grammatically: it's being documented as a constructor, not a conversion routine.  The timeout "is" the value in ticks.  But sure.

File: include/kernel.h
Comment (andyross) [reply include/kernel.h]: To conform with k_uptime_get() mostly, which was (probably -- it predates me, but I agree) done to prevent the application author from computing a time delta in unsigned units and tripping over the resulting overflow.

File: include/kernel.h
Comment (andyross) [reply include/kernel.h]: It really should not, but we can have that fight later :)

File: include/sys_clock.h
Comment (andyross) [reply include/sys_clock.h]: No, in legacy mode the value stored in the k_timeout_t is a count of milliseconds.

File: include/sys_clock.h
Comment (andyross) [reply include/sys_clock.h]: There's a big comment above it's implementation describing what it does, with a warning about how to use it correctly.  In general we don't doxygen internal functions.  It's declared here only so it can be used in an inline.

File: include/sys_clock.h
Comment (andyross) [reply include/sys_clock.h]: It's a little subtle in general, honestly, thus that comment.  I don't know that a simple name is going to really help that much -- this is an internal gadget used in like three places.

File: kernel/queue.c
Comment (andyross) [reply kernel/queue.c]: It was never needed.  The assumption seemed to be that k_poll() might fail spuriously and be retried, but k_poll() doesn't fail in a way that requires retries.  We should be propagating any internal errors, not hiding them.

And in this case fixing the design bug was easier than trying to refactor it to do retries over a timeout.

File: kernel/sched.c
Comment (andyross) [reply kernel/sched.c]: Mostly because I'm trying not to change the legacy mode code paths too much -- the code used to be done in the various subsystems above this function, so I left it that way.  The new code does, indeed, fold all the conversion code to a single place (the timeout constructor macros) and all the tick alignment to z_add_timeout().

File: kernel/timeout.c
Comment (andyross) [reply kernel/timeout.c]: Yes, but why?  In the new design, it just passes through a k_timeout_t that was constructed all the way up in application code without having to convert or inspect anything in the intervening layers.  That's better, I'd argue.

Comment (andyross) [reply kernel/timeout.c]: Good point.  I'm wondering why that collision isn't a warning....  Will fix.

File: kernel/timeout.c
Comment (andyross) [reply kernel/timeout.c]: Whoops, this was a rebase error.  Will fix.

Comment (andyross) [reply kernel/timeout.c]: Actually no, this is fine.  The lock got moved up a level so the caller could add its own math inside the lock.

File: drivers/i2s/Kconfig
Comment (nordic-krch) [reply drivers/i2s/Kconfig]: Fair enough, i guess that they will be attacked one by one to enable new timeout API.

File: samples/scheduler/metairq_dispatch/src/msgdev.c
Comment (nordic-krch) [reply samples/scheduler/metairq_dispatch/src/msgdev.c]: what i meant was that if `Z_TIMEOUT_TICKS` would be created for legacy then ifdef would not be needed. 

Currently, it is defined as `#define Z_TIMEOUT_TICKS(t) k_ticks_to_ms_ceil32(t)`. What if it is defined as `#define Z_TIMEOUT_TICKS(t) t`. That would simplify also use in `shed.c:1161` to something like: `timeout = Z_TIMEOUT_TICKS(ticks + (IF_ENABLED(LEGACY) ? _TICK_ALIGN : 0));`

File: include/sys_clock.h
Comment (nordic-krch) [reply include/sys_clock.h]: yes, but this macro is used in the context where `t` is ticks value:
https://github.com/andyross/zephyr/blob/10aaf24035586b315645d3cd9146ef52362cec68/kernel/sched.c#L1162
https://github.com/andyross/zephyr/blob/10aaf24035586b315645d3cd9146ef52362cec68/samples/scheduler/metairq_dispatch/src/msgdev.c#L78

File: include/sys_clock.h
Comment (nordic-krch) [reply include/sys_clock.h]: ok, i found the comment in the implementation. Imo, such comments should be in the header even though these functions are not public.

File: kernel/timeout.c
Comment (nordic-krch) [inline kernel/timeout.c]: seems to me that it should be `- (curr_tick + elapsed())`

File: kernel/sched.c
Comment (nordic-krch) [reply kernel/sched.c]: legacy will stay for long maybe it would be good to have it clean. Looking at places where z_add_timeout is used, it seems to me that it could take k_timeout_t for legacy code as well. There is only one place where `z_add_timeout` is called with ticks without earlier conversion from milliseconds: `msgdev.c` and it is a sample using internal api so i guess it could be solved somehow (e.g. force new timeout scheme only).

File: kernel/Kconfig
Comment (nordic-krch) [reply kernel/Kconfig]: yes, it should be `depends on ... + default y`

File: include/sys_clock.h
Comment (andyross) [reply include/sys_clock.h]: That's correct.  The final argument to z_add_timeout() has changed: in legacy code it's a 32 bit count of ticks (not milliseconds) which ipc mechanisms compute from application-provided arguments.  In new code it's a k_timeout_t that is constructed in the application code.

File: include/sys_clock.h
Comment (andyross) [reply include/sys_clock.h]: Yeah, I think we have to agree to disagree on this one.  I really don't think putting documentation for internal functions into public headers is a good idea.

File: kernel/sched.c
Comment (andyross) [reply kernel/sched.c]: I'd really prefer to do that cleanup later.  I really don't think keeping a few 2-5 line blocks of code unchanged under an #if is that much of a burden.  And the intent is that LEGACY goes away in a release or two anyway.

File: kernel/timeout.c
Comment (andyross) [reply kernel/timeout.c]: Arrgh, right as I was about to push.  Yes, this is a bug.  Will fix and update.

**Comment by andyross** [comment]
Rebased again around a minor collision.  Cleaned up the long line warnings from checkpatch.  Added the discussed doc changes and a few other minor things from review.  Fixed the elapsed() bug @nordic-krch  found.

Should be good to go.

File: kernel/sched.c
Comment (nordic-krch) [reply kernel/sched.c]: can you please add comment in `z_add_timeout` explaining why for legacy implementation value stored in `k_timeout_t` is not actual k_timeout_t but ticks. Someone else looking at that code may get confused (someone else can be me or you in few months :)).

**Comment by nordic-krch** [comment]
I think it would be good to align timeout parameter documentation in `kernel.h`

For example `k_thread_create` delay is described as `Scheduling delay, or K_NO_WAIT (for no delay)`.
but I noticed that here https://github.com/zephyrproject-rtos/zephyr/blob/698000f40c6e9f44fbc7753b259ae5e92ba90fc2/tests/kernel/threads/thread_apis/src/test_threads_spawn.c#L119 `k_thread_create` is called with `K_FOREVER`.

Some examples:
- `Non-negative waiting period to obtain a data item or one of the special values K_NO_WAIT and                K_FOREVER.`
- `Delay before submitting the work item` (no mention about special values)

Imo, they could all refer to `k_timeout_t`. Like:
- `Waiting period to obtain  a data item. See k_timeout_t for details.`
- `Delay before submitting the work item. See k_timeout_t for details.`
- etc.

**Comment by nordic-krch** [comment]
@andyross what about CI? Any chance to get results there?

**Comment by andyross** [comment]
Are those really something that should block merge?  The documentation as written is correct.  I agree it could be improved.  Can't we file an issue on that or just submit it personally post-merge?

As far as CI: no, PRs that touch this many tests don't complete.  Someone has to run it manually before merge.

**Comment by andyross** [comment]
Rebased

**Comment by nashif** [comment]
> @andyross what about CI? Any chance to get results there?

@nordic-krch given the amount of the changes CI times out here, this will go through a manual run before merge. If you see any build/test issues related to this change, please raise them. Approval of this PR should not depend on CI passing.

The only alterntive is to raise the number of shippable minions  in .shippable.yml and let CI complete and then remove the CI change before the merge. @andyross You can do that if you feel like it.

**Comment by nashif** [comment]
just noticed this one:

```
/home/buildslave/src/github.com/zephyrproject-rtos/zephyr/drivers/timer/xlnx_psttc_timer.c:164:12: error: invalid operands to binary == (have 's32_t' {aka 'int'} and 'k_timeout_t' {aka 'struct <anonymous>'})
  164 |  if (ticks == K_FOREVER) {
      |            ^~
```

**Comment by andyross** [comment]
Correct some rebase crosses in built-but-not-run platform and one spot in tests/kernel/context that wasn't compatible and literally just merged.

**Comment by nordic-krch** [comment]
there are 2 issues remaining:
- `loapic_timer.c`
- mcuboot. I've created PR for mcuboot: https://github.com/JuulLabs-OSS/mcuboot/pull/693
However, mcuboot part has some dependencies since #23625 required mcuboot changes which already got merged (https://github.com/JuulLabs-OSS/mcuboot/pull/682)

**Comment by andyross** [comment]
Fixed up the acrn failure (shouldn't acrn be built under sanitycheck if it's run under CI?  What broke there?)

Similarly: when did mcuboot move out of tree?  I don't even know how to run those tests, why are they in CI?

**Comment by nashif** [comment]
>     * mcuboot. I've created PR for mcuboot: [JuulLabs-OSS/mcuboot#693](https://github.com/JuulLabs-OSS/mcuboot/pull/693)

I do not think it is a good idea to be depending on upstream PRs, I would first submit this and make it work in the repo maintained in west and do the upstreaming in parallel, otherwise PRS will be stuck for a long time depending on how fast upstream reacts.

**Comment by andyross** [comment]
Rebase to resolve collisions with commit a4c919064973 and commit 84921c53b1b5 which snuck around this one into master.

Note I did run this through a full sanitycheck myself, just spot checked the relevant tests (kernel/threads and kernel/interrupt) kernel/common across the standard platforms.

**Comment by carlescufi** [comment]
> > ```
> > * mcuboot. I've created PR for mcuboot: [JuulLabs-OSS/mcuboot#693](https://github.com/JuulLabs-OSS/mcuboot/pull/693)
> > ```
> 
> I do not think it is a good idea to be depending on upstream PRs, I would first submit this and make it work in the repo maintained in west and do the upstreaming in parallel, otherwise PRS will be stuck for a long time depending on how fast upstream reacts.

What we've been doing for tightly-coupled projects like MCUboot is to send a PR upstream, point to that PR and remote in the `west.yml` in the zephyr PR, then merge the upstream MCUboot PR, then updated our MCUboot copy with the latest MCUboot, then point the `west.yml` in this PR to our MCUboot.
@nvlsianpu can help with that, or @nordic-krch or myself can do that

**Comment by andyross** [comment]
Er, excuse me: I did **not** run this through sanitycheck before pushing.

As far as that mcuboot PR, it shows merged now.  I'm hoping we're OK now.

**Comment by nashif** [comment]
DNM so we remember to remove the shippable hack.

**Review by nordic-krch — APPROVED** [review]

**Comment by andyross** [comment]
One more to clean up another crossed issue with the new cavs_timer driver.  This also includes a build fix to an upstream bug with cc26x2r1_launchxl  that is merely unmasked here.  It's not a timer driver thing at all, but it's a simple and obvious patch.

Note that we're still seeing a failure in mcuboot though.  @nordic-krch , can you take a look at that?

**Comment by andyross** [comment]
OK, we're good now.  Just the one mcuboot failure left.

**Comment by vanti** [comment]
Fyi the cc26x2r1_launchxl issue is being addressed in #23929.

**Comment by carlescufi** [comment]
@andyross please rebase, the mcuboot build should be fixed now

**Comment by nordic-krch** [comment]
Remaining issue with mcuboot in PR: https://github.com/JuulLabs-OSS/mcuboot/pull/700

**Comment by andyross** [comment]
Rebased (actually the only collision was that last patch to cc26x2r1_launchxl , which made mine unnecessary so I just removed it).

**Review by nashif — APPROVED** [review]

**Pull request merged**
